modules/images/webp-uploads/load.php

```diff
@@ -35,5 +35,186 @@ function webp_uploads_filter_image_editor_output_format( $output_format, $filena
+			}
+
+			$editor->resize( (int) $properties['width'], (int) $properties['height'], $properties['crop'] );
+			$filename = $editor->generate_filename( null, null, $extension );
+			$image    = $editor->save( $filename, $mime );
+
+			if ( is_wp_error( $image ) ) {
+				continue;
+			}
+
+			// TODO: Store the file size of the created image.
+			// $image['filesize'] = filesize( $image['path'] );
+			// Remove duplicated properties from the size image.
+			unset( $image['path'], $image['height'], $image['width'], $image['mime-type'] );
+
+			$sources[ $mime ] = $image;
+		}
+
+		$current_size['sources'] = $sources;
+		$sizes[ $size ]          = $current_size;
+	}
+
+	$metadata['sizes'] = $sizes;
+
+	return $metadata;
+}
+
+/**
+ * List all the available image sizes as well with the properties for each, the properties included
+ * are: width, height and crop. The logic behind this function tries to use the defined sizes by
+ * WordPress if the values is not found on the options it would fallback to the user defined
+ * sizes. Each property uses the default for each property `null` for height and width and `false`
+ * for crop the same way as `make_subsize` the editor image.
+ *
+ * @since n.e.x.t
+ *
+ * @see   WP_Image_Editor_Imagick::make_subsize()
+ * @see   WP_Image_Editor_GD::make_subsize()
+ * @return array An array with the details of all available image sizes: width, height and crop.
+ */
+function webp_uploads_get_image_sizes() {
+	$wp_image_sizes = wp_get_additional_image_sizes();
+	$sizes          = array();
+
+	// Create the full array with sizes and crop info.
+	foreach ( get_intermediate_image_sizes() as $size ) {
+		// Set the default values similar to `make_subsize`.
+		$width  = null;
+		$height = null;
+		$crop   = false;
+
+		if ( array_key_exists( $size, $wp_image_sizes ) ) {
+			if ( array_key_exists( 'width', $wp_image_sizes[ $size ] ) ) {
+				$width = (int) $wp_image_sizes[ $size ]['width'];
+			}
+			if ( array_key_exists( 'height', $wp_image_sizes[ $size ] ) ) {
+				$height = (int) $wp_image_sizes[ $size ]['height'];
+			}
+			if ( array_key_exists( 'crop', $wp_image_sizes[ $size ] ) ) {
+				$crop = (bool) $wp_image_sizes[ $size ]['crop'];
+			}
+		}
+
+		/**
+		 * Get the values from the option if is not present on the options' fallback to the
+		 * defined image sizes instead due if the option is not present it means is a custom image size.
+		 */
+		$sizes[ $size ] = array(
+			'width'  => get_option( $size . '_size_w', $width ),
+			'height' => get_option( $size . '_size_h', $height ),
+			'crop'   => (bool) get_option( $size . '_crop', $crop ),
+		);
+	}
+
+	return $sizes;
+}
+
+/**
+ * Return an array with the list of valid mime types for a sources' property.
+ *
+ * @since n.e.x.t
+ *
+ * @todo Add a filter to support more mime types.
+ *
+ * @return string[] An array of valid mime types, where the key is the mime type and the value is the extension type.
+ */
+function webp_uploads_valid_image_mime_types() {
```

Comment:
We should think about a better name for this function, it's not really about _valid_ image MIME types. For example, PNG is also a valid MIME type.

Also, I think it would be more appropriate to have this function return a _map_ of which image MIME types can be transformed to another image MIME types, since that's what we're actually looking for here. Storing the extensions here is duplicated information, we can already get this from WordPress core's `wp_get_mime_types()` function.

We need to make sure this function also makes sense for the future, and what we really care about here is which image sizes core should support generating versions in another MIME type for.

Because of the above, I think we should change this to a function that returns a map of supported MIME type "transformations", e.g.
```php
return array(
	'image/jpeg' => array( 'image/webp' ),
	// And potentially, if we also decide to support that:
	'image/webp' => array( 'image/jpeg' ),
);
```

That would easily allow to in the future also control which file types should be transformed to AVIF, or to also transform PNG images to WebP if that becomes possible.

Of course the logic above that consumes this function would then need to be adjusted: We could check whether the original MIME type is one of the keys in this "transformation" array, and then we generate sub-sizes for all MIME types that are in the value array.

The function could be called something like `webp_uploads_get_supported_image_mime_transforms()`.

Reply:
Agree, having a map makes more sense in the long term to support additional mime types for a defined mime type.